### PR TITLE
Allow to use QAOA-CV optimizer inside classification module

### PR DESCRIPTION
**Pull request by gcattan**
This PR allows to use QAOA-CV docplex optimizer inside the classification module.
To date, only naive QAOA optimizer and classical optimization are available.

Issues: https://github.com/pyRiemann/pyRiemann-qiskit/issues/29, https://github.com/pyRiemann/pyRiemann-qiskit/issues/30

### CODE REVIEW

File: pyriemann_qiskit/classification.py [hunk 1]
```diff
@@ -675,6 +710,13 @@ class QuanticMDM(QuanticClassifierBase):
     qaoa_optimizer : SciPyOptimizer, default=SLSQP()
         An instance of a scipy optimizer to find the optimal weights for the
         parametric circuit (ansatz).
+    create_mixer : Callable[int, QuantumCircuit], default=None
```
Comment (qbarthelemy) [inline pyriemann_qiskit/classification.py:713]:
```suggestion
    create_mixer : None | Callable[int, QuantumCircuit], default=None
```

File: pyriemann_qiskit/classification.py [hunk 2]
```diff
@@ -1059,6 +1113,13 @@ class QuanticNCH(QuanticClassifierBase):
...
```
Comment (qbarthelemy) [inline pyriemann_qiskit/classification.py:1116]:
```suggestion
    create_mixer : None | Callable[int, QuantumCircuit], default=None
```

File: pyriemann_qiskit/utils/docplex.py [hunk 3]
```diff
@@ -11,13 +11,15 @@
 from docplex.mp.vartype import ContinuousVarType, IntegerVarType, BinaryVarType
 import numpy as np
 from pyriemann.utils.covariance import normalize
+from pyriemann_qiskit.utils.hyper_params_factory import create_mixer_rotational_X_gates
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:14]: Can you move these imports below, as:
```
from .hyper_params_factory import create_mixer_rotational_X_gates
from .math import is_pauli_identity
```

File: pyriemann_qiskit/utils/docplex.py [hunk 4]
```diff
@@ -645,17 +660,85 @@ def prepare_model(qp):
...
+    Notes
+    -----
+    .. versionadded:: 0.4.0
+
+    References
+    ----------
+    .. [1] \
+        http://ibmdecisionoptimization.github.io/docplex-doc/mp/_modules/docplex/mp/model.html#Model
+
+    """
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:701]: What class or function does this documentation correspond to?

File: pyriemann_qiskit/utils/docplex.py [hunk 5]
```diff
@@ -645,17 +660,85 @@ def prepare_model(qp):
...
+    -------
+    docplex_weights : dict
+        A vector of integer decision variables representing
+        our weights.
+
+    Notes
+    -----
+    .. versionadded:: 0.4.0
+
+    """
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:726]: What class or function does this documentation correspond to?

File: pyriemann_qiskit/utils/math.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 from typing_extensions import deprecated
 
 import numpy as np
+from qiskit.quantum_info import Pauli, SparsePauliOp
 from pyriemann.utils.covariance import normalize
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/math.py:6]: Imports must be sorted in alphabetic order:
```suggestion
from pyriemann.utils.covariance import normalize
from qiskit.quantum_info import Pauli, SparsePauliOp
```

**Review by qbarthelemy — CHANGES_REQUESTED** [review]

File: pyriemann_qiskit/utils/docplex.py (hunk 4 above)
Comment (gcattan) [reply pyriemann_qiskit/utils/docplex.py:701]: This is a link to the `docplex.mp.model` implementation. 
May be not too much useful after all.

Comment (gcattan) [reply pyriemann_qiskit/utils/docplex.py:701]: It is referenced here: 
image
Do you want me to remove it?

**Review by gcattan — COMMENTED** [review]
rename imports

File: pyriemann_qiskit/utils/docplex.py (hunk 4 above)
Comment (qbarthelemy) [reply pyriemann_qiskit/utils/docplex.py:701]: No. My point is that I don't understand what class (or function) this docstring documents.

Normally, it is
```
def new_func(arg0, arg1):
    """ HERE, DOCUMENTATION """
    do_stuff
    return smart_result
```

Comment (gcattan) [reply pyriemann_qiskit/utils/docplex.py]: Oh, ok.

File: pyriemann_qiskit/utils/docplex.py [hunk 7]
```diff
@@ -5,23 +5,25 @@
 - MDM optimization problem;
 - computation of matrices mean.
 """
+import numpy as np
 import math
 import time
 
 from docplex.mp.vartype import ContinuousVarType, IntegerVarType, BinaryVarType
-import numpy as np
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:12]: No, numpy must be imported here, not in Python standard libs.

File: pyriemann_qiskit/utils/docplex.py [hunk 8]
```diff
@@ -201,114 +203,110 @@ class pyQiskitOptimizer:
...
+            such as channels'name or number but None.
+        name : string
+            An custom name for the variable. The name is used internally by docplex
+            and may appear if your print the model to a file for example.
+
+        Returns
+        -------
+        docplex_covmat : dict
+            A square matrix of decision variables representing
+            our covariance matrix.
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:244]: "our" is a bit weird in a doc.

File: pyriemann_qiskit/utils/docplex.py [hunk 9]
```diff
@@ -201,114 +203,110 @@ class pyQiskitOptimizer:
...
+        prob : Model
+            An instance of the docplex model [1]_
+        classes : list
+            The classes.
+
+        Returns
+        -------
+        docplex_weights : dict
+            A vector of decision variables representing
+            our weights.
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:303]: "our", same remark.

File: pyriemann_qiskit/utils/docplex.py [hunk 10]
```diff
@@ -201,114 +203,110 @@ class pyQiskitOptimizer:
...
-    -----
-    .. versionadded:: 0.0.2
-    """
-
     def convert_covmat(self, covmat):
+        """Hook to apply some transformation on a covariance matrix.
+
+        Parameters
+        ----------
+        covmat : ndarray, shape (n_features, n_features)
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:211]: Like in pyRiemann, try to replace `covmat` inputs by `X`, more `sklearn`istic.

And try to generalize covariance matrices to SPD matrices.

File: pyriemann_qiskit/utils/docplex.py [hunk 11]
```diff
@@ -201,114 +203,110 @@ class pyQiskitOptimizer:
...
-    """
-
     def covmat_var(self, prob, channels, name):
+        """Helper to create a docplex representation of a
+        covariance matrix variable.
+
+        Parameters
+        ----------
+        prob : Model
+            An instance of the docplex model [1]_
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:232]:
```suggestion
            An instance of the docplex model [1]_.
```

File: pyriemann_qiskit/utils/docplex.py [hunk 12]
```diff
@@ -343,42 +341,41 @@ def __init__(self, optimizer=CobylaOptimizer(rhobeg=2.1, rhoend=0.000001)):
...
+
+        Notes
+        -----
+        .. versionadded:: 0.0.2
+
+        References
+        ----------
+        .. [1] \
+            http://ibmdecisionoptimization.github.io/docplex-doc/mp/_modules/docplex/mp/model.html#Model
+
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:377]:
```suggestion
```

File: pyriemann_qiskit/utils/docplex.py [hunk 13]
```diff
@@ -388,36 +385,49 @@ def _solve_qp(self, qp, reshape=True):
...
+        prob : Model
+            An instance of the docplex model [1]_
+        classes : list
+            The classes.
+
+        Returns
+        -------
+        docplex_weights : dict
+            A vector of continuous decision variables representing
+            our weights.
```
Comment (qbarthelemy) [inline pyriemann_qiskit/utils/docplex.py:403]: "our"

**Review by qbarthelemy — CHANGES_REQUESTED** [review]

**Review by qbarthelemy — CHANGES_REQUESTED** [review]

File: benchmarks/light_benchmark.py [hunk 14]
```diff
@@ -22,7 +22,9 @@
 from moabb import set_log_level
```
Comment (qbarthelemy) [inline benchmarks/light_benchmark.py:22]: Can you sort all these imports?

File: benchmarks/light_benchmark.py [hunk 15]
```diff
@@ -11,25 +11,29 @@
...
+from pyriemann_qiskit.utils.hyper_params_factory import create_mixer_rotational_X_gates
+from qiskit_algorithms.optimizers import SPSA
+from sklearn.decomposition import PCA
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
+from sklearn.metrics import balanced_accuracy_score
+from sklearn.model_selection import train_test_split
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import LabelEncoder
 import sys
+import warnings
```
Comment (qbarthelemy) [inline benchmarks/light_benchmark.py:35]: `sys` and `warnings` are standard libraries, and must be imported in the first block of imports.

**Review by qbarthelemy — APPROVED** [review]
Thx @gcattan !

**Comment by gcattan** [comment]
Thank you for the review :)

**Pull request merged**